aep/general/0122/aep.md

```diff
@@ -0,0 +1,318 @@
+# Resource paths
+
+Most APIs expose _resources_ (their primary nouns) which users are able to
+create, retrieve, and manipulate. Additionally, resources have _paths_: each
+resource has a unique identifying path that users use to reference that
+resource, and these paths are what users should _store_ as the canonical path
+for the resources.
+
+## Guidance
+
+All resource paths defined by an API **must** be unique within that API. (See
+the section on [full resource paths](#full-resource-paths) below for more
+information on referring to resources across APIs.)
+
+Resource paths are formatted according to the [URI path schema][], but without
+the leading slash:
+
+    publishers/123/books/les-miserables
+    users/vhugo1802
+
+- Resource path components **should** usually alternate between collection
+  identifiers (example: `publishers`, `books`, `users`) and resource IDs
+  (example: `123`, `les-miserables`, `vhugo1802`).
+- Resource paths **must** use the `/` character to separate individual segments
+  of the resource path.
+  - Non-terminal segments of a resource path **must not** contain a `/`
+    character.
+  - The terminal segment of a resource path **should not** contain a `/`
+    character.
+- Resource paths **should** only use characters available in DNS names, as
+  defined by [RFC-1123](https://tools.ietf.org/html/rfc1123).
+  - Additionally, resource IDs **should not** use upper-case letters.
+  - If additional characters are necessary, resource paths **should not** use
+    characters that require URL-escaping, or characters outside of ASCII.
+  - If Unicode characters can not be avoided, resource paths **must** be stored
+    in Normalization Form C (see [AIP-210][]).
+- Eacg resource **must** expose a `path` field that contains its resource path.
+  - Resources **may** provide the resource ID, i.e. the last segment of the
+    path, as a separate field named `id`.
+  - Resources **may** expose a separate, system-generated unique ID field
+    (`uid`).
+  - Resources **must not** expose tuples, self-links, or other forms of
+    resource identification.
+  - All ID fields **must** be strings.
+
+**Note:** Resource paths as described here are used within the scope of a
+single API (or else in situations where the owning API is clear from the
+context), and are only required to be unique within that scope. For this
+reason, they are sometimes called _relative resource paths_ to distinguish them
+from _full resource paths_ (discussed below).
+
+[aip-210]: ./0210.md
+[uri path schema]: https://datatracker.ietf.org/doc/html/rfc3986#appendix-A
+
+### Collection identifiers
+
+The collection identifier segments in a resource path **must** be the plural
+form of the noun used for the resource. (For example, a collection of
+`Publisher` resources is called `publishers` in the resource path.)
+
+- Collection identifiers **must** be concise American English terms.
```

Comment:
I agree it's hard to lint (though maybe not in our post-LLM reality)...this is what human API design review is for.

I do a lot of API design reviews, and 95% of it is stuff about the product details being underspecified or the goals being unclear, but 5% of it is noticing stuff like this which the API linter can't catch.  Given that the API design reviews are needed anyway, I don't think it's a big problem.

The AIPs started, IIUC, as an authoritative and canonical reference for human API design reviews, rather than as a spec for generic clients.  I don't object to a more client-heavy focus, but I don't think we need to scour the human parts.

Reply:
Discussed live:  leave as is (with some reluctance)